File: code/24.swap-nodes-in-pairs.go
/*
 * @lc app=leetcode id=24 lang=golang
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{
		Next: head,
	}
	cur := dummy

	for cur.Next != nil && cur.Next.Next != nil {
		temp := cur.Next
		next := cur.Next.Next.Next
		cur.Next = cur.Next.Next
		cur.Next.Next = temp
		cur.Next.Next.Next = next
		cur = cur.Next.Next
	}
	return dummy.Next
}

// @lc code=end

File: code/24.swap-nodes-in-pairs.java
/*
 * @lc app=leetcode id=24 lang=java
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start

import java.util.ListIterator;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode next = head.next;
        ListNode node = swapPairs(next.next);

        head.next = node;
        next.next = head;
        return next;
    }
}
// @lc code=end
